clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -511,17 +511,18 @@ private static int parseAcks(String acksString) {
      * @return The amount of time we waited in ms
      */
     private long waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
-        // add topic to metadata topic list if it is not there already.
-        if (!this.metadata.containsTopic(topic))
-            this.metadata.add(topic);
-
+        // add topic to metadata topic list if it is not there already and reset expiry
+        this.metadata.add(topic);
         if (metadata.fetch().partitionsForTopic(topic) != null)
             return 0;
 
         long begin = time.milliseconds();
         long remainingWaitMs = maxWaitMs;
         while (metadata.fetch().partitionsForTopic(topic) == null) {
             log.trace("Requesting metadata update for topic {}.", topic);
+            // Add the topic to metadata before requesting update to ensure that the topic is included
+            // in the metadata topic list since unknown topics are removed from the metadata
+            this.metadata.add(topic);
```

Comment:
@ijuma @hachikuji I have changed this PR to address only topic expiry. Have created another JIRA  (https://issues.apache.org/jira/browse/KAFKA-3792) to reduce log spam for non-existent topics. I think consumer should timeout as well similar to the producer instead of waiting forever, but since there is no consumer config to do that currently, more discussion is required. Will wait to see what @hachikuji thinks.